File: algorithms/sorting/bubble_sort.py
# Bubble Sort

not_sorted_array = [2, 8, 5, 3, 9, 4, 1]
print("Before:   ", not_sorted_array)

# Time Complexity:      O(n ** 2)
# Space Complexity :    O(1)
def bubble_sort_v1(array):
    n = len(array)

    for i in range(n - 1):
        for j in range(n - 1 - i):
            if array[j] > array[j + 1]:
                array[j], array[j + 1] = array[j + 1], array[j]

    return None

not_sorted_array_v1 = not_sorted_array.copy()
bubble_sort_v1(not_sorted_array_v1)

print("After V1: ", not_sorted_array_v1)


# Time Complexity:      θ(n ** 2), Ω(n)
# Space Complexity:    O(1)
def bubble_sort_v2(array):
    # We did some optimization of with checking if array is already done
    n = len(array)

    for i in range(n - 1):
        swapped = False

        for j in range(n - 1 - i):
            if array[j] > array[j + 1]:
                array[j], array[j + 1] = array[j + 1], array[j]
                swapped = True

        if not swapped:
            break

    return None

not_sorted_array_v2 = not_sorted_array.copy()
bubble_sort_v2(not_sorted_array_v2)

print("After V2: ", not_sorted_array_v2)

File: cpp_algorithms/sorting/bubble_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

// Bubble Sort V1
//
// Time Complexity:
// O(n^2)
//
// Space Complexity:
// O(1)
void bubble_sort_v1(vector<int>& array) {
    int n = array.size();

    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - 1 - i; j++) {
            if (array[j] > array[j + 1]) {
                swap(array[j], array[j + 1]);
            }
        }
    }
}


// Bubble Sort V2
//
// Time Complexity:
// Worst/Average: O(n^2)
// Best:          O(n)
//
// Space Complexity:
// O(1)
void bubble_sort_v2(vector<int>& array) {
    int n = array.size();

    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;

        for (int j = 0; j < n - 1 - i; j++) {
            if (array[j] > array[j + 1]) {
                swap(array[j], array[j + 1]);
                swapped = true;
            }
        }

        // Array is already sorted
        if (!swapped) {
            break;
        }
    }
}


void print_array(const vector<int>& array) {
    for (int x : array) {
        cout << x << ' ';
    }
    cout << '\n';
}


int main() {
    vector<int> not_sorted_array = {2, 8, 5, 3, 9, 4, 1};

    cout << "Before:   ";
    print_array(not_sorted_array);

    vector<int> not_sorted_array_v1 = not_sorted_array;

    bubble_sort_v1(not_sorted_array_v1);

    cout << "After V1: ";
    print_array(not_sorted_array_v1);


    vector<int> not_sorted_array_v2 = not_sorted_array;

    bubble_sort_v2(not_sorted_array_v2);

    cout << "After V2: ";
    print_array(not_sorted_array_v2);

    return 0;
}
